File: backend/src/foragerr/ws/messages.py
"""Domain-event → WebSocket push-message mapping (FRG-API-010).

Every message the endpoint broadcasts is the SignalR-shaped envelope the
frontend bridge consumes (``frontend/src/ws/messages.ts``)::

    {"name": <resource family>, "action": <verb>, "resource": <payload>}

The bus publishes concrete domain events; :func:`map_event` is the single
place they become that wire envelope. Events with no mapping return ``None``
and are dropped, so subscribing to the base ``object`` type on the bus (to
catch every family without enumerating them at subscribe time) stays safe.

Divergence from the scaffold's queue-progress contract (reported, not hidden)
-----------------------------------------------------------------------------
``frontend/src/ws/messages.ts`` models a ``{name:'queue', action:'progress'}``
message carrying ``{id, page, progress, sizeLeft, status}`` and PATCHES the
cached page in place. The backend cannot produce that message from the events
it actually emits:

* :class:`TrackedStateChanged` fires only on a state TRANSITION, not on every
  byte-count tick, and carries the string ``download_id`` — not the numeric
  ``tracked_downloads.id`` the queue rows are keyed by, and no ``size``/
  ``sizeleft``/``progress`` fields.
* ``page`` is a React-Query pagination concept the backend has no knowledge of.

So queue changes are emitted as ``{name:'queue', action:'updated'}`` — an
INVALIDATION signal — and the frontend must refetch the queue query on it
rather than patch. The scaffold bridge only patches (action ``progress``) or
invalidates ``series``; it needs a queue-invalidation branch to consume these.
Live byte-level progress patching is deferred (needs a per-tick event carrying
the numeric id + sizes; M2).
"""

from __future__ import annotations

from typing import Any

from foragerr.commands.service import CommandStatusChanged
from foragerr.downloads.tracking import DownloadFailedEvent, TrackedStateChanged
from foragerr.library.flows import SeriesRefreshed

#: (name, action, resource) — the three fields of the wire envelope.
ResourceMessage = tuple[str, str, dict[str, Any]]


def map_event(event: Any) -> ResourceMessage | None:
    """Translate a published domain event into a push envelope, or ``None``.

    Coverage (FRG-API-010): queue (tracked-download state + failures), series
    (metadata refresh), command (status change). Issue-file changes have no
    backend emitter in M1 (design note: M1 scopes broadcast to queue+command);
    the mapping is trivially extensible when such an event lands.
    """
    if isinstance(event, TrackedStateChanged):
        # `status` here is the lifecycle state string the frontend calls
        # `status`; `health` is the ok/warning/error rollup (backend `status`).
        return (
            "queue",
            "updated",
            {
                "downloadId": event.download_id,
                "status": event.state,
                "health": event.status,
                "seriesId": event.series_id,
                "issueId": event.issue_id,
            },
        )
    if isinstance(event, DownloadFailedEvent):
        return (
            "queue",
            "updated",
            {
                "downloadId": event.download_id,
                "status": "failed",
                "issues": [list(pair) for pair in event.issues],
            },
        )
    if isinstance(event, SeriesRefreshed):
        return ("series", "updated", {"id": event.series_id, "partial": event.partial})
    if isinstance(event, CommandStatusChanged):
        return (
            "command",
            "updated",
            {"id": event.id, "name": event.name, "status": event.status},
        )
    return None


__all__ = ["ResourceMessage", "map_event"]

File: frontend/src/ws/messages.ts
import type { QueueItem } from '../api/types';

/*
 * WebSocket push message shape (FRG-UI-001): {name, action, resource}.
 *   name     — resource family, e.g. 'series' | 'queue'
 *   action   — 'updated' | 'progress' | ...
 *   resource — the payload (a partial resource or progress record)
 */
export interface WsMessage {
  name: string;
  action: string;
  resource: unknown;
}

/**
 * Progress payload carried by a `{name:'queue', action:'progress'}` message.
 * `progress`/`sizeLeft` are OPTIONAL: a tick may carry only a status transition,
 * in which case the bridge preserves the row's existing numeric values.
 */
export interface QueueProgressResource {
  id: number;
  page: number;
  progress?: number;
  sizeLeft?: number;
  status?: QueueItem['status'];
}

export function isQueueProgress(msg: WsMessage): msg is WsMessage & {
  resource: QueueProgressResource;
} {
  if (msg.name !== 'queue' || msg.action !== 'progress') return false;
  const r = msg.resource as Partial<QueueProgressResource> | null;
  if (!r || typeof r.id !== 'number' || typeof r.page !== 'number') return false;
  // When present, progress/sizeLeft MUST be numbers — otherwise a malformed
  // tick would render "undefined%" once patched into the cached row.
  if (r.progress !== undefined && typeof r.progress !== 'number') return false;
  if (r.sizeLeft !== undefined && typeof r.sizeLeft !== 'number') return false;
  return true;
}

/**
 * Command lifecycle payload carried by a `{name:'command', action:'updated'}`
 * message (backend/src/foragerr/ws/messages.py maps `CommandStatusChanged` to
 * `{id, name, status}`). All optional here because the bridge only reads it to
 * classify a message and must never throw on an unexpected shape.
 */
export interface CommandResource {
  id?: number;
  name?: string;
  status?: string;
}

/** The pull-refresh command name (backend `PULL_REFRESH_TRIGGERED_BY`). */
export const PULL_REFRESH_COMMAND = 'pull-refresh';

/**
 * True only for a `command` message that reports the pull-refresh command
 * reaching `completed` — the single command transition that rewrites this
 * week's stored pull entries. The backend pushes a `command` message on EVERY
 * lifecycle transition of EVERY command, so the bridge must NOT invalidate the
 * (expensive, whole-week) pull projection on all of them — only this one.
 */
export function isPullRefreshComplete(msg: WsMessage): boolean {
  if (msg.name !== 'command') return false;
  const r = msg.resource as Partial<CommandResource> | null;
  return !!r && r.name === PULL_REFRESH_COMMAND && r.status === 'completed';
}

export function parseWsMessage(raw: string): WsMessage | null {
  try {
    const parsed = JSON.parse(raw) as Partial<WsMessage>;
    if (typeof parsed.name === 'string' && typeof parsed.action === 'string') {
      return { name: parsed.name, action: parsed.action, resource: parsed.resource };
    }
    return null;
  } catch {
    return null;
  }
}
